collections/_posts/2022-10-03-MomentumResidualNeuralNetwork.md

```diff
@@ -0,0 +1,151 @@
+
+$$v_n=\frac{1}{\gamma} (v_{n+1}-(1-\gamma)f(x_n,\theta_n))$$
+
+This avoids the memory issues occuring due to the backpropagation step.
+
+Note: Momentum gradient descent is an alternative to classical gradient descent algorithms using second order partial derivatives.
+A overview of gradient descent optimization algorithms is given in the paper of S.Ruder (http://arxiv.org/abs/1609.04747) as mentioned in the paper.
+
+## Memory cost
+For usual ResNets, one need to store the weights of the network and the values of all activations for the training set at each layer. The memory needed is $$O(k*d*n_{batch})$$ while for the Momentum ResNets the memory need is $$O((1-\gamma)*k*d*n_{batch})$$.
+
+![](/collections/images/MomentumResNets/fig_memory_momentum.jpg)
+
+## The role of momentum
+When $$\gamma=0$$, they get a Classical ResNet. When $$\gamma \rightarrow 1$$, they get a special case of the invertible RevNet (Gomez et al, 2017). 
+
+The advantage of the Momentum ResNet compared to RevNet where two learnable functions are used, is its stability for convergence (proofs given in the paper).
+
+## Link with ODE (Ordinary Differential Equations)
+
+![](/collections/images/MomentumResNets/fig_momentum_odes.jpg)
+ 
+ResNets can be interpreted as a first order ODE. Indeed the term $$x_{n+1} - x_n$$ can be seen as the discretized partial derivative $$\dot{x}$$.
+
+$$ \dot{x} \quad  \rightarrow  \quad x_{n+1}-x_{n}$$
+
+Momentum ResNets can be interpreted as second-order ODEs by taking $$\epsilon=\frac{1}{1-\gamma}$$, they get:
+
+$$ v_{n+1}=v_{n} + \frac{f(x_n,\theta_n) - v_n}{\epsilon} $$
+
+$$ x_{n+1}=x_n+v_{n+1}$$
+
+
+by replacing the term $$v_n$$, they find:
+
+$$x_{n+1}-x_{n}=x_n-x_{n-1} + \frac{f(x_n,\theta_n)-v_n}{\epsilon}$$
+
+The second order derivative $$\ddot{x}$$ can be discretized using $$x_{n+1}-2x_n+x_{n-1}$$ and then we can interpret the Momentum ResNets as a second order ODE of the form:
+
+$$\epsilon \ddot{x}+\dot{x}=f(x,\theta)$$
+
+Briefly : "In the same way that ResNets can be seen as discretization of first order ODEs, Momentum ResNets can be seen as discretization of second-order ones."
+
+When $$\epsilon \rightarrow 0$$, they get the first order model.
+
+## Representation capabilies
+These analogies between ODEs lead to some interesting mathematical properties. The first order model can represent homeomorphism mappings (continuous, bijective with continuous inverse). However first order ODEs are not universal approximators and some mappings are not possible (see the example on point clouds separation). Momentum ResNets can capture non-homeomorphic dynamics. The authors present some proofs on this aspect called "representation capabilities" of models.
+
+# Experiments
+
+## Point clouds separation
+
+* Experiments on 4 rings (2 classes) of point clouds 
+![](/collections/images/MomentumResNets/momentum_points_clouds.jpg)
+
+These classes failed to be separated using classical ResNets but not with Momentum ResNets. 
```

Comment:
I think a clearer way to phrase this would be:
```suggestion
Momentum ResNets are able to separate these classes, whereas classical ResNets fail.
```